### ISSUE TITLE
Local Relay Model (nostrdb)

### ISSUE THREAD

**Issue by jb55** [damus-io/damus#2041]
To switch to a local query model, we need to move away from trusting remote relays.

There are many issues with what we are currently doing:

- We don't check to see if the results returned from the relay actually match the query we sent. This enables rogue relays to inject ads or other bad data. This is particularily bad in the gossip model, since users can force clients to make requests to evil relays outside of the user's configured relay pool.

- Since we don't check signatures on all note types (only profiles), this would enable users to spoof DMs. This is technically possible today but is mitigated by some level of trust in the current mega-relays.

- We have to spin off async processes parse note contents once we get them off the wire. Swift's concurrency model is decent but it is still kind of annoying to delay timeline insertion based off a bunch of asyncronous tasks.

- Since we're currently inserting data into nostrdb already we technically are doing twice the amount of processing work for no reason 

Switching to a nostrdb for local queries fixes all of these issues:

- All notes in nostrdb are validated, so we don't have to worry about spoofing

- We don't have to worry about relays sending bad data, because we only ever query the local relay which we know will always match the query we are executing.

- Local subscriptions are notified only after we parse note contents, so we don't have to wait for async processing to render the note.

- nostrdb is much more efficient at ingesting notes, so we will save CPU and battery.

- Having notes locally will allow us to use negentropy for keeping nostrdb nodes in sync.

This is already implemented in Damus notedeck and android, now we need to overhaul Damus iOS to support this model. This issue is for tracking this!

Once we switch to this model we can begin work on adding support for the gossip model!

**Comment by alltheseas** [comment]
@jb55 to use this as iOS nostrdb tracking issue

**Comment by alltheseas** [comment]
@jb55 commits to "thursday & friday from now on to this"

**Comment by jb55** [comment]
> [@jb55](https://github.com/jb55) commits to "thursday & friday from now on to this"

well that was a lie

**Comment by alltheseas** [comment]
@daniel advised two (2) story points for draft

**Comment by danieldaquino** [comment]
Created a sub-issue for the draft and moved the story points there.

**Comment by alltheseas** [comment]
It appears hashtag search in universe view (e.g. type in #foodstr, #vianegativa etc.) is in perpetual load state

<img width="250" height="512" alt="Image" src="https://github.com/user-attachments/assets/9c33791b-1e26-4a5f-9250-dcb254eccacd" />

@danieldaquino

**Comment by alltheseas** [comment]
Mute list is borked for Will and me

**Comment by alltheseas** [comment]
From Vanessa

> i'm on the new testflight - feed not updating and notifications showing ones from 1 year/2 years ago

**Comment by alltheseas** [comment]
Stale feed persists

**Comment by danieldaquino** [comment]
> It appears hashtag search in universe view (e.g. type in #foodstr, #vianegativa etc.) is in perpetual load state
> 
> <img alt="Image" width="250" height="512" src="https://private-user-images.githubusercontent.com/64376233/486151260-9c33791b-1e26-4a5f-9250-dcb254eccacd.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTc1MTkwMTMsIm5iZiI6MTc1NzUxODcxMywicGF0aCI6Ii82NDM3NjIzMy80ODYxNTEyNjAtOWMzMzc5MWItMWUyNi00YTVmLTkyNTAtZGNiMjU0ZWNjYWNkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA5MTAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwOTEwVDE1MzgzM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWVkNDk5ZDk2OTg0NGE0ZDZjNDZjZGZlYWQxMWQ3ZDRiNzg5ZGJlMGIxYzk0OWFhN2MyM2Y5ZWZmM2M2YmU4OTYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.u01b-eEZGkBs5NLx_cEg6G49KU40-H1roMir-cztDm8">
> [@danieldaquino](https://github.com/danieldaquino)

@alltheseas can you please confirm which version you saw this on?

**Comment by alltheseas** [comment]
Hashtag perpetual load is fixed on 1.15 (1131) a855739. 

The perpetual load report is from a version six (6) days ago. 

Just re-installed 1128, and recreated perpetual hashtag load

![image](https://github.com/user-attachments/assets/6e360be4-83e0-4da4-a9a8-770556c78fdc)

**Comment by alltheseas** [comment]
@danieldaquino on mute being broken what I observe is:

1) when I try to add a new muted profile, damus iOS advises I don't have a mute list,
2) and asks me to overwrite / start a new list
3) if I click yes, the profile is muted
4) if I mute another profile, I am presented with step 1)

thereby continuously resetting my mute list

**Comment by alltheseas** [comment]
@danieldaquino - @jb55 submitted a background crash. I also submitted a few crashes past 24 hours.

**Comment by danieldaquino** [comment]
> [@danieldaquino](https://github.com/danieldaquino) - [@jb55](https://github.com/jb55) submitted a background crash. I also submitted a few crashes past 24 hours.

Thank you @alltheseas, I have analyzed the crash logs received.

I believe those new crashes are related to the app lifecycle. There seems to be two similar kinds of crashes happening:
1. This one is related to the app holding onto the NostrDB database file after going to the background
2. This one is related to the app accessing NostrDB memory after it has been closed.

I am investigating how I can fix those and improve the app background/foreground signal handling logic.

**Comment by alltheseas** [comment]
Noticed today on latest internal TF 1.15 (1143) 33aa214
whenever I post, quote, re-post my post does not appear immediately. If I close the app, and re-open to event appears.

Maybe related to stale feed?

**Comment by danieldaquino** [comment]
> Noticed today on latest internal TF 1.15 (1143) [33aa214](https://github.com/damus-io/damus/commit/33aa2141efba5cdc4926076eea2a1d71df4ae3b9) whenever I post, quote, re-post my post does not appear immediately. If I close the app, and re-open to event appears.
> 
> Maybe related to stale feed?

Yes, I believe that would be a side-effect of the stale timeline issue.

**Comment by alltheseas** [comment]
Replies are very slow to load


https://github.com/user-attachments/assets/ce492dba-890a-4a4e-a923-601d92f7a56d

**Comment by alltheseas** [comment]
Stale feed is resolved on latest internal testflight.

> threads are taking much longer to load on experimental

https://github.com/user-attachments/assets/4fd51868-b726-46aa-9c73-3d73e3a33336

> 3 second thread load is pretty consistent

observed by @jb55 and me

**Comment by alltheseas** [comment]
see https://github.com/damus-io/damus/pull/3280 as prospective solution to "nil" error in #3279 
`
Unexpectedly found nil” (damus/ContentView.swift:305)`
